model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java

```diff
@@ -0,0 +1,136 @@
+
+/**
+ *
+ * @author hmlnarik
+ */
+public class ConcurrentHashMapStorage<K, V extends AbstractEntity<K>, M> implements MapStorage<K, V, M> {
+
+    private final ConcurrentMap<K, V> store = new ConcurrentHashMap<>();
+
+    private final Map<SearchableModelField<M>, UpdatePredicatesFunc<K, V, M>> fieldPredicates;
+
+    @SuppressWarnings("unchecked")
+    public ConcurrentHashMapStorage(Class<M> modelClass) {
+        this.fieldPredicates = MapFieldPredicates.getPredicates(modelClass);
+    }
+
+    @Override
+    public V create(K key, V value) {
+        return store.putIfAbsent(key, value);
+    }
+
+    @Override
+    public V read(K key) {
+        Objects.requireNonNull(key, "Key must be non-null");
+        return store.get(key);
+    }
+
+    @Override
+    public V update(K key, V value) {
+        return store.replace(key, value);
+    }
+
+    @Override
+    public boolean delete(K key) {
+        return store.remove(key) != null;
+    }
+
+    @Override
+    public long delete(ModelCriteriaBuilder<M> criteria) {
+        long res;
+        if (criteria == null) {
+            res = store.size();
+            store.clear();
+            return res;
+        }
+
+        MapModelCriteriaBuilder<K, V, M> b = criteria.unwrap(MapModelCriteriaBuilder.class);
+        if (b == null) {
+            throw new IllegalStateException("Incompatible class: " + criteria.getClass());
+        }
+        Predicate<? super K> keyFilter = b.getKeyFilter();
+        Predicate<? super V> entityFilter = b.getEntityFilter();
+        res = 0;
+        for (Iterator<Entry<K, V>> iterator = store.entrySet().iterator(); iterator.hasNext();) {
+            Entry<K, V> next = iterator.next();
+            if (keyFilter.test(next.getKey()) && entityFilter.test(next.getValue())) {
+                res++;
+                iterator.remove();
+            }
+        }
+        return res;
+    }
+
+    @Override
+    public ModelCriteriaBuilder<M> createCriteriaBuilder() {
+        return new MapModelCriteriaBuilder<>(fieldPredicates);
+    }
+
+    @Override
+    public MapKeycloakTransaction<K, V, M> createTransaction() {
+        return new MapKeycloakTransaction<>(this);
+    }
+
+
+    @Override
+    public Stream<V> read(ModelCriteriaBuilder<M> criteria) {
+        if (criteria == null) {
+            return Stream.empty();
+        }
+        Stream<Entry<K, V>> stream = store.entrySet().stream();
+
+        MapModelCriteriaBuilder<K, V, M> b = criteria.unwrap(MapModelCriteriaBuilder.class);
+        if (b == null) {
+            throw new IllegalStateException("Incompatible class: " + criteria.getClass());
+        }
+        Predicate<? super K> keyFilter = b.getKeyFilter();
+        Predicate<? super V> entityFilter = b.getEntityFilter();
+        stream = stream.filter(me -> keyFilter.test(me.getKey()) && entityFilter.test(me.getValue()));
+
+        return stream.map(Map.Entry::getValue);
+    }
+
+    public Stream<V> read() {
```

Comment:
Good catch. This method should have been removed.
Reason for not including it in the interface is to remove easy way to read full database which is not desirable anywhere apart from test and development where it can be achieved via the existing `read(ModelCriteriaBuilder)` method.